nav2_waypoint_follower/plugins/photo_at_waypoint.cpp

```diff
@@ -138,7 +138,7 @@ bool PhotoAtWaypoint::processAtWaypoint(
 void PhotoAtWaypoint::imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
 {
   std::lock_guard<std::mutex> guard(global_mutex_);
-  curr_frame_msg_ = msg;
+  curr_frame_msg_ = std::make_shared<sensor_msgs::msg::Image>(*msg);
```

Comment:
Please explain why storing shared pointers of shared pointer stored data. This should be OK.

Reply:
My concern is that the shared object is owned by the Inters-Process Manager, not by the PhotoAtWaypoint node.
The data in the object can be modified outside of the callback.

Reply:
Huh?

Reply:
most of my changes are protecting against the `msg` object being modified outside the callback.
if you assume that can't happen, then the code is fine as-was.

Reply:
That can't happen in the case that I'm mentioning above. I didn't check all of them, but I suspect many of them fall into that same category. The executor is single threaded so if this is processing, that promises that another cannot be

Reply:
it happens using `rmw_fastrtps_cpp` as discussed in the other threads.

Reply:
^ if not in Cyclone (so rclcpp bug isn't impacting you) we can revert these since it'll be fixed in the next sync

Reply:
Should it not be sufficient to change the callback signature to `void PhotoAtWaypoint::imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr msg)`?

Reply:
From other conversations in other threads, no, I think not. I think that makes it const for us in the function, but the middleware doesn't have that const-ness